compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/PromisedValue.kt

```diff
@@ -98,17 +101,60 @@ abstract class BooleanValue(codegen: ExpressionCodegen) :
     }
 }
 
-class BooleanConstant(codegen: ExpressionCodegen, val value: Boolean) : BooleanValue(codegen) {
-    override fun jumpIfFalse(target: Label) = if (value) Unit else mv.goTo(target)
-    override fun jumpIfTrue(target: Label) = if (value) mv.goTo(target) else Unit
+class ConstantValue(codegen: ExpressionCodegen, val value: Any?, irType: IrType) : BooleanValue(codegen, value.asmType(), irType) {
     override fun materializeAt(target: Type, irTarget: IrType) {
-        mv.iconst(if (value) 1 else 0)
-        if (Type.BOOLEAN_TYPE != target) {
-            StackValue.coerce(Type.BOOLEAN_TYPE, target, mv)
+        when (value) {
+            is Boolean -> mv.iconst(if (value) 1 else 0)
+            is Char -> mv.iconst(value.toInt())
+            is Long -> mv.lconst(value)
+            is Float -> mv.fconst(value)
+            is Double -> mv.dconst(value)
+            is Number -> mv.iconst(value.toInt())
+            else -> mv.aconst(value)
         }
+        // `null` is of any reference type.
+        if (value != null || AsmUtil.isPrimitive(target)) {
+            coerceOnStack(target, irTarget)
+        }
+    }
+
+    override fun discard() {
+        // Add something for the line number to be attached to.
+        mv.nop()
     }
 
-    override fun discard() {}
+    override fun jumpIfFalse(target: Label) = if (isDefaultValueForType(type, value)) mv.goTo(target) else mv.nop()
```

Comment:
ConstanValue class  and especially `jumpIfXXXX` method a bit confusing, cause only boolean constant should be branched  values. Maybe add assertion about types here? 
Or refactor values similar to old backend approached, where all constants except boolean are StackValue.Constant and boolean constants are BranchedValue.Companion.TRUE/FALSE

Reply:
Technically any value can be converted to a branched value via `coerceToBoolean`, and treating 0s as false is consistent with that.

Reply:
Technically yes, but could such situation arised in bytecode generation?

Reply:
With generics yes, but with constants no. Removed this inheritance for non-boolean constants.